### DISCUSSION TITLE
running on-prem

### DISCUSSION THREAD

**Discussion by LYDE1234** [microsoft/rest-api-fuzz-testing#148] [General]
We have a custom development automated testing framework that is used for extensive functional testing of a mature client / server product. There's ongoing work to build REST APIs for the existing interfaces, and we'd like to extend automated testing to that, potentially including fuzzing. I realize this project is intended for Azure deployment but is it possible to deploy and use on-prem for what's described above, and how would you go about doing it?

**Comment by mgreisen** [comment]
@LYDE1234 it would depend on your infrastructure. If you wanted to replicate the complete system, you would need to provide a frontend service to injest jobs, a message passing service, and an orchestrator that would deploy jobs into an isolated target (a VM for example). This really depends on your resources you have available and your requirements. 

We are in the process of adding a feature for local deployment of jobs that would not require azure. This means you could deploy a job on your own machine or on a build machine within a pipeline for example. The only dependencies would be python and docker. 

This first implementation will have some limitations. For example if you need to deploy the service locally as well as the test tools, you will only be able to deploy one at a time. However, if your test tools are running against services that are deployed elsewhere, i.e. some kind of pre-production test environment, you can deploy as many test jobs as you like. 

See this branch https://github.com/stishkin/rest-api-fuzz-testing-pub/blob/local-docker/cli/raft_local.py
We are hoping to have this committed soon.

**Comment by mgreisen** [comment]
@LYDE1234 We've released the raft_local.py script in our recent v3.0 release.

**Reply by bottiger** [reply]
Hi!

This is great news. However, could you please help out by adding some documentation on requirements and how to get the local version running?

**Reply by mgreisen** [reply]
@bottiger I've just added a document that should help you get started. Look at the file how-to-use-raft-local.md in the docs directory.

**Reply by bottiger** [reply]
@mgreisen Thanks a lot for the quick response! I will look at it asap at let you know if I run into any problems with it. However, so far it seems good enough for getting me started
